File: behavioral-design-pattern/chain-of-responsibility-pattern/chain_of_responsibility.go
// Chain of Responsibility Pattern
// ================================
// Category: Behavioral Design Pattern
//
// Intent:
//   Pass a request along a chain of handlers. Each handler either processes
//   the request or forwards it to the next handler.
//
// When to use:
//   - HTTP middleware, event processing, support escalation
//   - When the handler isn't known in advance
//   - When handlers should be configurable dynamically
//
// Key Participants:
//   - Handler (interface): Declares Handle and SetNext
//   - ConcreteHandler: Handles specific requests, forwards the rest
//   - Client: Sends request to the first handler

package main

import "fmt"

// ---------------------------------------------------------------------------
// Request data
// ---------------------------------------------------------------------------

type Priority int

const (
	Low      Priority = 1
	Medium   Priority = 2
	High     Priority = 3
	Critical Priority = 4
)

type SupportTicket struct {
	ID          string
	Title       string
	Priority    Priority
	Description string
	HandledBy   string
}

// ---------------------------------------------------------------------------
// Handler interface
// ---------------------------------------------------------------------------

type SupportHandler interface {
	SetNext(handler SupportHandler) SupportHandler
	Handle(ticket *SupportTicket)
}

// ---------------------------------------------------------------------------
// Base handler with next-chaining logic (embedded by concrete handlers)
// ---------------------------------------------------------------------------

type BaseHandler struct {
	next SupportHandler
}

func (b *BaseHandler) SetNext(handler SupportHandler) SupportHandler {
	b.next = handler
	return handler
}

func (b *BaseHandler) HandleNext(ticket *SupportTicket) {
	if b.next != nil {
		b.next.Handle(ticket)
	} else {
		fmt.Printf("  [Unhandled] Ticket %s reached end of chain\n", ticket.ID)
	}
}

// ---------------------------------------------------------------------------
// Concrete Handlers
// ---------------------------------------------------------------------------

type AutoResponder struct{ BaseHandler }

func (h *AutoResponder) Handle(ticket *SupportTicket) {
	if ticket.Priority == Low {
		ticket.HandledBy = "AutoResponder (Bot)"
		fmt.Printf("  [AutoResponder] Ticket %s: Sent automated FAQ response\n", ticket.ID)
	} else {
		h.HandleNext(ticket)
	}
}

type Level1Support struct{ BaseHandler }

func (h *Level1Support) Handle(ticket *SupportTicket) {
	if ticket.Priority == Medium {
		ticket.HandledBy = "Level 1 Support Agent"
		fmt.Printf("  [Level1] Ticket %s: Assigned to support agent\n", ticket.ID)
	} else {
		h.HandleNext(ticket)
	}
}

type Level2Support struct{ BaseHandler }

func (h *Level2Support) Handle(ticket *SupportTicket) {
	if ticket.Priority == High {
		ticket.HandledBy = "Level 2 Senior Engineer"
		fmt.Printf("  [Level2] Ticket %s: Escalated to senior engineer\n", ticket.ID)
	} else {
		h.HandleNext(ticket)
	}
}

type ManagerSupport struct{ BaseHandler }

func (h *ManagerSupport) Handle(ticket *SupportTicket) {
	if ticket.Priority == Critical {
		ticket.HandledBy = "Support Manager"
		fmt.Printf("  [Manager] Ticket %s: CRITICAL — Manager notified\n", ticket.ID)
	} else {
		h.HandleNext(ticket)
	}
}

// ---------------------------------------------------------------------------
// Usage Example
// ---------------------------------------------------------------------------
func main() {
	auto := &AutoResponder{}
	l1 := &Level1Support{}
	l2 := &Level2Support{}
	mgr := &ManagerSupport{}

	// Build the chain
	auto.SetNext(l1).SetNext(l2).SetNext(mgr)

	tickets := []*SupportTicket{
		{ID: "T-001", Title: "Password reset", Priority: Low},
		{ID: "T-002", Title: "App crashes", Priority: Medium},
		{ID: "T-003", Title: "Data corruption", Priority: High},
		{ID: "T-004", Title: "Service outage", Priority: Critical},
	}

	fmt.Println("Processing support tickets through the chain:\n")
	for _, ticket := range tickets {
		auto.Handle(ticket)
		fmt.Printf("    -> Handled by: %s\n\n", ticket.HandledBy)
	}
}

File: behavioral-design-pattern/chain-of-responsibility-pattern/chain_of_responsibility.rs
//! Chain of Responsibility Pattern
//! ================================
//! Category: Behavioral Design Pattern
//!
//! Intent:
//!   Pass a request along a chain of handlers. Each handler either processes
//!   the request or forwards it to the next.
//!
//! When to use:
//!   - Middleware pipelines, event processing, escalation systems
//!   - When the handler isn't known in advance
//!
//! Rust approach:
//!   We use a Vec of boxed handler closures rather than linked objects,
//!   which is more idiomatic in Rust and avoids lifetime complexity.

// ---------------------------------------------------------------------------
// Request data
// ---------------------------------------------------------------------------
#[derive(Debug, Clone, PartialEq)]
enum Priority { Low, Medium, High, Critical }

#[derive(Debug)]
struct SupportTicket {
    id: String,
    title: String,
    priority: Priority,
    handled_by: Option<String>,
}

// ---------------------------------------------------------------------------
// Handler trait
// ---------------------------------------------------------------------------
trait SupportHandler {
    fn can_handle(&self, ticket: &SupportTicket) -> bool;
    fn process(&self, ticket: &mut SupportTicket);
    fn name(&self) -> &str;
}

// ---------------------------------------------------------------------------
// Concrete Handlers
// ---------------------------------------------------------------------------
struct AutoResponder;
impl SupportHandler for AutoResponder {
    fn can_handle(&self, ticket: &SupportTicket) -> bool {
        ticket.priority == Priority::Low
    }
    fn process(&self, ticket: &mut SupportTicket) {
        ticket.handled_by = Some("AutoResponder (Bot)".to_string());
        println!("  [AutoResponder] Ticket {}: Sent automated FAQ response", ticket.id);
    }
    fn name(&self) -> &str { "AutoResponder" }
}

struct Level1Support;
impl SupportHandler for Level1Support {
    fn can_handle(&self, ticket: &SupportTicket) -> bool {
        ticket.priority == Priority::Medium
    }
    fn process(&self, ticket: &mut SupportTicket) {
        ticket.handled_by = Some("Level 1 Support Agent".to_string());
        println!("  [Level1] Ticket {}: Assigned to support agent", ticket.id);
    }
    fn name(&self) -> &str { "Level1" }
}

struct Level2Support;
impl SupportHandler for Level2Support {
    fn can_handle(&self, ticket: &SupportTicket) -> bool {
        ticket.priority == Priority::High
    }
    fn process(&self, ticket: &mut SupportTicket) {
        ticket.handled_by = Some("Level 2 Senior Engineer".to_string());
        println!("  [Level2] Ticket {}: Escalated to senior engineer", ticket.id);
    }
    fn name(&self) -> &str { "Level2" }
}

struct ManagerSupport;
impl SupportHandler for ManagerSupport {
    fn can_handle(&self, ticket: &SupportTicket) -> bool {
        ticket.priority == Priority::Critical
    }
    fn process(&self, ticket: &mut SupportTicket) {
        ticket.handled_by = Some("Support Manager".to_string());
        println!("  [Manager] Ticket {}: CRITICAL — Manager notified", ticket.id);
    }
    fn name(&self) -> &str { "Manager" }
}

// ---------------------------------------------------------------------------
// Chain — iterates through handlers until one processes the ticket
// ---------------------------------------------------------------------------
struct SupportChain {
    handlers: Vec<Box<dyn SupportHandler>>,
}

impl SupportChain {
    fn new() -> Self {
        Self { handlers: Vec::new() }
    }

    fn add(mut self, handler: Box<dyn SupportHandler>) -> Self {
        self.handlers.push(handler);
        self
    }

    fn handle(&self, ticket: &mut SupportTicket) {
        for handler in &self.handlers {
            if handler.can_handle(ticket) {
                handler.process(ticket);
                return;
            }
        }
        println!("  [Unhandled] Ticket {} reached end of chain", ticket.id);
    }
}

// ---------------------------------------------------------------------------
// Usage Example
// ---------------------------------------------------------------------------
fn main() {
    let chain = SupportChain::new()
        .add(Box::new(AutoResponder))
        .add(Box::new(Level1Support))
        .add(Box::new(Level2Support))
        .add(Box::new(ManagerSupport));

    let mut tickets = vec![
        SupportTicket { id: "T-001".into(), title: "Password reset".into(), priority: Priority::Low, handled_by: None },
        SupportTicket { id: "T-002".into(), title: "App crashes".into(), priority: Priority::Medium, handled_by: None },
        SupportTicket { id: "T-003".into(), title: "Data corruption".into(), priority: Priority::High, handled_by: None },
        SupportTicket { id: "T-004".into(), title: "Service outage".into(), priority: Priority::Critical, handled_by: None },
    ];

    println!("Processing support tickets through the chain:\n");
    for ticket in &mut tickets {
        chain.handle(ticket);
        println!("    -> Handled by: {}\n", ticket.handled_by.as_deref().unwrap_or("Nobody"));
    }
}
